drafts/source_generators_mvvm_toolkit.md

```diff
@@ -0,0 +1,160 @@
+---
+title: Source Generators in the MVVM Community Toolkit
+domain: software-engineering-corner.hashnode.dev
+tags: dotnet,mvvm,xaml,maui,csharp
+cover: https://cdn.hashnode.com/res/hashnode/image/upload/v1697099583332/Lv8fEMIab.png?auto=format
+publishAs: NiZarper
+hideFromHashnodeCommunity: false
+---
+
+# Source Generators in the MVVM Community Toolkit
+
+This blog post is a continuation of the blog post [Reducing boilerplate code with the MVVM Community Toolkit in .NET MAUI](https://software-engineering-corner.hashnode.dev/reducing-boilerplate-code-with-the-mvvm-community-toolkit-in-net-maui). It covers the Source Generator aspect of the MVVM Community Toolkit. Firstly, I am going to explain you the fundaments with some theory. Afterwards you are going to see an example from the MVVM Community Toolkit. The final chapter is going to summarize the blog post.
+
+## Table Of Contents
+
+- Introduction to Source Generators
+- Implementation of a Source Generator in the MVVM Community Toolkit
+- Takeaways
+
+### Introduction to Source Generators
+
+This chapter covers the basics of Source Generators and compares it to Runtime Reflection.
+Source Generators are a compiler feature which got released with .NET 5. Source Generators allow you to analyse compiled code and generate code based on it. You can also use files like CSV, JSON, XML, etc., as sources for code generation. Typically, Source Generators analyse the code for attributes or other conventions in the compiled code. The analysis happens via the Roslyn analysis APIs. The exciting part is that the generation occurs before the main code gets compiled. This means you can use the generated code inside the main code without any runtime overhead.
+
+The following image illustrates the Soure Generator step.
+
+<img src="https://cdn.hashnode.com/res/hashnode/image/upload/v1696929088614/v1LId0lWu.png?auto=format"/>
+
+[1] Source Generator step
+
+Typical usages for Source Generators are creating data classes from data files like CSV, JSON or XML and reducing boiler plate code, for instance, in XAML based solutions which use the interface INotifyPropertyChanged.
+It is important to mention that currently, Source Generators can only be implemented as <b>.NET Standard 2.0 libraries</b>.
+
+
+Before the introduction of Source Generators, generating code based on compiled code was already possible. This was done via Runtime Reflection.
+ASP.NET uses this extensively in e.g. controller classes, etc. The difference between Source Generators and Reflection is the point of time when the generation of code happens and the performance impact. For example, ASP.NET uses Runtime Reflection, there the generation of code occurs during the start-up. This means you can only accept requests once all the Runtime Reflection code is finished. Compared to Source Generators, where the code generation happens during the compile time. This would reduce the start-up time of an ASP.NET application.
```

Comment:
```suggestion
ASP.NET uses this extensively in e.g. controller classes, etc. The difference between Source Generators and Reflection is the point of time when the generation of code happens and the performance impact. For example, ASP.NET uses Runtime Reflection, where the generation of code occurs during the start-up. This means you can only accept requests once all the Runtime Reflection code is finished. Compared to Source Generators, where the code generation happens during the compile time. This would reduce the start-up time of an ASP.NET application.
```

Reply:
Also in the last sentence of this paragraph you say `This would reduce...`. Can you say that it actually does reduce the start-up time?

Reply:
If you look at it from a theoretical standpoint, yes. But there is certainly a reason why the ASP.NET team haven't changed the implementation until now.